src/app/components/artist_details/artist_details_model.rs

```diff
@@ -105,11 +109,68 @@ impl PlaylistModel for ArtistDetailsModel {
         )
     }
 
-    fn actions_for(&self, _: String) -> Option<gio::ActionGroup> {
-        None
+    fn actions_for(&self, id: String) -> Option<gio::ActionGroup> {
+        let songs = self.tracks_ref()?;
+        let song = songs.iter().find(|song| song.id == id)?;
+
+        let group = SimpleActionGroup::new();
+
+        let album_id = song.album.id.clone();
+        let view_album = SimpleAction::new("view_album", None);
+        let dispatcher = self.dispatcher.box_clone();
+        view_album.connect_activate(move |_, _| {
+            dispatcher.dispatch(AppAction::ViewAlbum(album_id.clone()));
+        });
+
+        group.add_action(&view_album);
+
+        for (i, artist) in song.artists.iter().enumerate() {
+            let view_artist = SimpleAction::new(&format!("view_artist_{}", i), None);
+            let dispatcher = self.dispatcher.box_clone();
+            let id = artist.id.clone();
+            view_artist.connect_activate(move |_, _| {
+                dispatcher.dispatch(AppAction::ViewArtist(id.clone()));
+            });
+            group.add_action(&view_artist);
+        }
+
+        let track_id = song.id.clone();
+        let copy_uri = SimpleAction::new("copy_uri", None);
+        copy_uri.connect_activate(move |_, _| {
+            let clipboard = Clipboard::get(&SELECTION_CLIPBOARD);
+            clipboard.set_text(&format!("spotify:track:{}", track_id.clone()));
+        });
+        group.add_action(&copy_uri);
+
+        let track_id = song.id.clone();
+        let copy_link = SimpleAction::new("copy_link", None);
+        copy_link.connect_activate(move |_, _| {
+            let clipboard = Clipboard::get(&SELECTION_CLIPBOARD);
+            clipboard.set_text(&format!(
+                "https://open.spotify.com/track/{}",
+                track_id.clone()
+            ));
+        });
+        group.add_action(&copy_link);
+
+        Some(group.upcast())
     }
 
-    fn menu_for(&self, _: String) -> Option<gio::MenuModel> {
-        None
+    fn menu_for(&self, id: String) -> Option<gio::MenuModel> {
+        let songs = self.tracks_ref()?;
+        let song = songs.iter().find(|song| song.id == id)?;
+
+        let menu = gio::Menu::new();
+        menu.insert(0, Some("View album"), Some("song.view_album"));
+        for (i, artist) in song.artists.iter().enumerate() {
+            menu.insert(
+                (i + 1) as i32,
```

Comment:
I had not realized there was an append method, my bad :facepalm: would you mind using append here?